File: python/src/threads/agents/tool.py
"""`tool()`: an app tool whose input type is a Pydantic model (spec/api.json `tool`).

The model is the tool's one schema: its JSON Schema is the pinned `input_schema` the model sees,
and the same model parses the arguments at the boundary (strict, JSON mode). There is no second
validator to drift from it.
"""

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Literal, Required, TypedDict, Unpack

from pydantic import BaseModel, JsonValue, TypeAdapter, ValidationError
from pydantic_core import to_json

from threads.agents.config import ConfigError
from threads.agents.context import RunContext
from threads.log import EffectClass, JsonObject, ToolSpec
from threads.log.jcs import canonicalize
from threads.loop.model import Found, LookupResult, NotFound, NotFoundNonfinal
from threads.loop.tools import Output
from threads.result import Ok

_OBJECT: TypeAdapter[dict[str, JsonValue]] = TypeAdapter(dict[str, JsonValue])


@dataclass(frozen=True, slots=True)
class Reconcile[O, D]:
    """How a reconcilable tool finds out whether its effect happened."""

    lookup: Callable[[str, RunContext[D]], Awaitable[LookupResult[O]]]
    finality: Literal["final", "nonfinal"]


class ToolOptions[I: BaseModel, O, D](TypedDict, total=False):
    name: Required[str]
    description: Required[str]
    input: Required[type[I]]
    runs: Literal["host", "sandbox"]
    execute: Callable[[I, RunContext[D]], Awaitable[O]]
    effect: EffectClass
    dedup_window_ms: int
    reconcile: Reconcile[O, D]
    ends_turn: bool


@dataclass(frozen=True, slots=True)
class Tool[I: BaseModel, O, D]:
    """An app tool (spec/api.json `Tool`). Build it with `tool()`."""

    name: str
    description: str
    input: type[I]
    execute: Callable[[I, RunContext[D]], Awaitable[O]]
    effect: EffectClass = "unguarded"
    dedup_window_ms: int | None = None
    reconcile: Reconcile[O, D] | None = None
    ends_turn: bool = False

    def spec(self) -> ToolSpec:
        """The pinned ToolSpec: what line 0 shows and what decides the effect class."""
        data: dict[str, JsonValue] = {
            "name": self.name,
            "description": self.description,
            "input_schema": json_schema(self.input),
            "effect_class": self.effect,
        }
        if self.dedup_window_ms is not None:
            data["dedup_window_ms"] = self.dedup_window_ms
        if self.ends_turn:
            data["ends_turn"] = True
        return ToolSpec.model_validate(data)

    def invalid(self, input: JsonObject) -> str | None:
        """Why the model's arguments fail the input model, or None."""
        why = invalid(self.input, dict(input))
        return None if why is None else f"invalid arguments for {self.name}: {why}"

    async def run(self, input: JsonObject, ctx: RunContext[D]) -> Output:
        """Runs the body. Its value is the result; an error it raises is an error result the
        model sees (spec/api.json `tool.execute`)."""
        text = canonicalize(dict(input))
        if not isinstance(text, Ok):
            return Output("the arguments are not canonical JSON", is_error=True)
        parsed = self.input.model_validate_json(text.value, strict=True)
        try:
            value = await self.execute(parsed, ctx)
        except Exception as error:
            # The tool's own failure is a result the model sees, not a crash of the run.
            return Output(f"{type(error).__name__}: {error}", is_error=True)
        return Output(render_value(value))

    async def lookup(self, effect_key: str, ctx: RunContext[D]) -> LookupResult[str]:
        """Reconciliation through the tool's own lookup; a `found` value becomes result text."""
        if self.reconcile is None:
            raise AssertionError("only a reconcilable tool is looked up")
        answer = await self.reconcile.lookup(effect_key, ctx)
        if isinstance(answer, Found):
            return Found(render_value(answer.value))
        if isinstance(answer, NotFound) and self.reconcile.finality == "nonfinal":
            # Finality is the tool's declared capability, never inferred from an answer.
            return NotFoundNonfinal()
        return answer


def json_schema(model: type[BaseModel]) -> dict[str, JsonValue]:
    """The model's JSON Schema as pinned: a tool's input_schema, an agent's output schema."""
    return _OBJECT.validate_python(model.model_json_schema())


def invalid(model: type[BaseModel], value: JsonValue) -> str | None:
    """Why a value the model sent fails `model`, or None. Strict and in JSON mode: `"1"` is not
    an int. One check for tool arguments and final_output candidates alike."""
    text = canonicalize(value)
    if not isinstance(text, Ok):
        return "not canonical JSON"
    try:
        model.model_validate_json(text.value, strict=True)
    except ValidationError as error:
        return f"{error.error_count()} error(s): {error}"
    return None


def render_value(value: object) -> str:
    """A tool's value as the text the model sees: a string as is, anything else as JSON."""
    if isinstance(value, str):
        return value
    if isinstance(value, BaseModel):
        return value.model_dump_json()
    return to_json(value).decode("utf-8")


def tool[I: BaseModel, O, D](**options: Unpack[ToolOptions[I, O, D]]) -> Tool[I, O, D]:
    """spec/api.json `tool`. `runs` defaults to "host" and is not pinned. Raises ConfigError for
    a definition that can't run: a sandbox tool (no sandbox in this build), an idempotent tool
    without its dedup window, or a reconcilable tool without its lookup."""
    effect = options.get("effect", "unguarded")
    execute = options.get("execute")
    if options.get("runs", "host") != "host" or execute is None:
        raise ConfigError("capability_missing", f"{options['name']}: only runs='host' with execute")
    window = options.get("dedup_window_ms")
    if (effect == "idempotent") != (window is not None):
        raise ConfigError("invalid_config", "dedup_window_ms goes with effect='idempotent' only")
    reconcile = options.get("reconcile")
    if (effect == "reconcilable") != (reconcile is not None):
        raise ConfigError("invalid_config", "reconcile goes with effect='reconcilable' only")
    return Tool(
        options["name"],
        options["description"],
        options["input"],
        execute,
        effect,
        window,
        reconcile,
        options.get("ends_turn", False),
    )

File: typescript/packages/core/src/agent/tool.ts
import { z } from "zod";
import {
  canonicalize,
  type EffectClass,
  JsonObject,
  JsonValue,
  type Principal,
  ToolSpec,
} from "../log";
import type { ToolImpl, ToolRun } from "../loop";
import type { LookupResult } from "../model";
import { ConfigError } from "./errors";

// tool() (spec/api.json): an app tool. Its Zod input schema is both the JSON Schema the model
// sees and the parser its arguments go through.

/** What tools and hooks receive (spec/api.json RunContext). Deps are host-only. */
export type RunContext<Deps = undefined> = {
  readonly deps: Deps | undefined;
  readonly threadId: string;
  readonly branchId: string;
  readonly principal: Principal;
  readonly callId?: string;
  /** `<branch_id>:<call_id>`: send it to providers that dedup. */
  readonly effectKey?: string;
  readonly signal: AbortSignal;
};

export type ToolDefinition<Input, Output, Deps = undefined> = {
  readonly name: string;
  readonly description: string;
  readonly input: z.ZodType<Input>;
  readonly output?: z.ZodType<Output>;
  /** Defaults to "host". "sandbox" is reserved: setup fails with capability_missing. Not pinned. */
  readonly runs?: "host" | "sandbox";
  readonly execute?: (input: Input, ctx: RunContext<Deps>) => Promise<Output>;
  /**. Undeclared tools are unguarded: uncertainty always parks. */
  readonly effect?: z.infer<typeof EffectClass>;
  readonly dedupWindowMs?: number;
  readonly reconcile?: {
    readonly lookup: (
      effectKey: string,
      ctx: RunContext<Deps>,
    ) => Promise<LookupResult<Output>>;
    readonly finality: "final" | "nonfinal";
  };
  readonly endsTurn?: boolean;
  /** true: runs with the other concurrent read-only calls of one response. Hashed, not in line 0. */
  readonly concurrent?: boolean;
  /** true: the model sees only its name, listed in tool_search's description, until tool_search loads it. */
  readonly defer?: boolean;
};

/** The environment a run binds its tools to. */
export type ToolEnv<Deps> = {
  readonly deps: Deps | undefined;
  readonly threadId: string;
  readonly branchId: string;
  readonly principal: Principal;
};

/** An app tool. `types` only carries Input and Output for inference. */
export type Tool<Input, Output, Deps = undefined> = {
  readonly name: string;
  readonly types?: { readonly input: Input; readonly output: Output };
  /** The pinned spec; throws ConfigError when the definition can't run. */
  readonly spec: () => ToolSpec;
  readonly bind: (env: ToolEnv<Deps>) => ToolImpl;
  /** Declared `concurrent: true`; pinned by config_hash as concurrent_tools. */
  readonly concurrent?: true;
  /** Declared `defer: true`: pinned in reference form when context.defer_tools is auto. */
  readonly defer?: true;
};

export function tool<Input, Output, Deps = undefined>(
  def: ToolDefinition<Input, Output, Deps>,
): Tool<Input, Output, Deps> {
  // Arguments must match the pinned schema exactly: an object schema refuses unknown keys.
  const strict: z.ZodType =
    def.input instanceof z.ZodObject ? def.input.strict() : def.input;
  const spec = (): ToolSpec => pinned(def, strict);
  return {
    name: def.name,
    spec,
    bind: (env) => bound(def, strict, spec(), env),
    ...(def.concurrent === true ? { concurrent: true } : {}),
    ...(def.defer === true ? { defer: true } : {}),
  };
}

function pinned<Input, Output, Deps>(
  def: ToolDefinition<Input, Output, Deps>,
  input: z.ZodType,
): ToolSpec {
  const effect = def.effect ?? "unguarded";
  if ((def.runs ?? "host") !== "host" || def.execute === undefined)
    throw new ConfigError(
      "capability_missing",
      `tool ${def.name}: only runs "host" with execute is supported in this release`,
    );
  if ((effect === "idempotent") !== (def.dedupWindowMs !== undefined))
    throw new ConfigError(
      "invalid_config",
      `tool ${def.name}: dedupWindowMs is required exactly when effect is idempotent`,
    );
  if ((effect === "reconcilable") !== (def.reconcile !== undefined))
    throw new ConfigError(
      "invalid_config",
      `tool ${def.name}: reconcile is required exactly when effect is reconcilable`,
    );
  if (
    def.concurrent === true &&
    (effect !== "read_only" || def.endsTurn === true)
  )
    throw new ConfigError(
      "invalid_config",
      `tool ${def.name}: concurrent needs effect: "read_only" and no endsTurn; tools with side effects, or that end the turn, run one at a time`,
    );
  if (def.defer === true && def.endsTurn === true)
    throw new ConfigError(
      "invalid_config",
      `tool ${def.name}: defer can't be combined with endsTurn; a deferred tool is loaded by tool_search first`,
    );
  const parsed = ToolSpec.safeParse({
    name: def.name,
    description: def.description,
    input_schema: jsonSchema(def.name, input),
    effect_class: effect,
    ...(def.dedupWindowMs === undefined
      ? {}
      : { dedup_window_ms: def.dedupWindowMs }),
    ...(def.output === undefined
      ? {}
      : { output_schema: jsonSchema(def.name, def.output) }),
    ...(def.endsTurn === true ? { ends_turn: true } : {}),
  });
  if (!parsed.success)
    throw new ConfigError(
      "invalid_config",
      `tool ${def.name}: ${z.prettifyError(parsed.error)}`,
    );
  return parsed.data;
}

/** The exported JSON Schema: the one source the model sees (AGENTS.md, Design idea). */
export function jsonSchema(
  name: string,
  schema: z.ZodType,
): z.infer<typeof JsonObject> {
  let exported: unknown;
  try {
    exported = z.toJSONSchema(schema, { io: "input" });
  } catch (error) {
    throw new ConfigError("invalid_config", `${name}: ${String(error)}`);
  }
  const { $schema: _dialect, ...rest } = JsonObject.parse(exported);
  return rest;
}

function bound<Input, Output, Deps>(
  def: ToolDefinition<Input, Output, Deps>,
  strict: z.ZodType,
  spec: ToolSpec,
  env: ToolEnv<Deps>,
): ToolImpl {
  const context = (
    callId: string | undefined,
    effectKey: string | undefined,
    signal: AbortSignal,
  ): RunContext<Deps> => ({
    ...env,
    signal,
    ...(callId === undefined ? {} : { callId }),
    ...(effectKey === undefined ? {} : { effectKey }),
  });
  const run = async (
    args: unknown,
    ctx: {
      readonly callId: string;
      readonly effectKey: string;
      readonly signal: AbortSignal;
    },
  ): Promise<ToolRun> => {
    const execute = def.execute;
    if (execute === undefined)
      throw new Error(`tool ${def.name} has no execute`);
    // The loop parsed the arguments with `strict` already; this gives execute its typed Input.
    const parsed = def.input.parse(args);
    try {
      return done(
        await execute(parsed, context(ctx.callId, ctx.effectKey, ctx.signal)),
        def.output,
      );
    } catch (error) {
      // spec/api.json choice tool-errors: a thrown error is a result the model sees.
      return { kind: "done", output: String(error), isError: true };
    }
  };
  const reconcile = def.reconcile;
  return {
    spec,
    input: strict,
    run,
    ...(def.concurrent === true ? { concurrent: true } : {}),
    ...(reconcile === undefined
      ? {}
      : {
          reconcile: {
            finality: reconcile.finality,
            lookup: async (key: string): Promise<LookupResult<string>> => {
              const answer = await reconcile.lookup(
                key,
                context(undefined, key, new AbortController().signal),
              );
              if (answer.status !== "found") return answer;
              const shown = done(answer.value, def.output);
              return shown.kind === "done" && !shown.isError
                ? { status: "found", value: shown.output }
                : {
                    status: "unknown",
                    reason: "the found value fails the output schema",
                  };
            },
          },
        }),
  };
}

/** The model-visible text of a value: a string as is, anything else as canonical JSON. */
function done<Output>(
  value: Output,
  schema: z.ZodType<Output> | undefined,
): ToolRun {
  if (schema !== undefined && !schema.safeParse(value).success)
    return {
      kind: "done",
      output: "the result fails the tool's output schema",
      isError: true,
    };
  if (typeof value === "string")
    return { kind: "done", output: value, isError: false };
  const json = JsonValue.safeParse(value);
  const text = json.success ? canonicalize(json.data) : undefined;
  return text?.ok === true
    ? { kind: "done", output: text.value, isError: false }
    : { kind: "done", output: "the result is not JSON", isError: true };
}
